src/Strings.f90

```diff
@@ -320,79 +315,263 @@ MODULE Strings
+
+    IF((pos > 0) .AND. (pos <= LEN(this))) THEN
+      letter = this%s(pos:pos)
+    ELSE
+      ALLOCATE(CHARACTER(0) :: letter)
+    ENDIF
+  ENDFUNCTION at_string
+!
+!-------------------------------------------------------------------------------
+!> @brief Returns the characters in the slice indicated, if the either index is
+!> out of range, then the ascii null character is returned
+!> @param this the string being sliced
+!> @param stt the first index of the slice
+!> @param stp the last index of the slice
+!
+  PURE FUNCTION at_string_slice(this,stt,stp) RESULT(slice)
+    CLASS(StringType),INTENT(IN) :: this
+    INTEGER(SIK),INTENT(IN) :: stt
+    INTEGER(SIK),INTENT(IN) :: stp
+    CHARACTER(LEN=:),ALLOCATABLE :: slice
+
+    IF((stt <= stp) .AND. (stt > 0) .AND. (stp <= LEN(this))) THEN
+      slice = this%s(stt:stp)
+    ELSE
+      ALLOCATE(CHARACTER(0) :: slice)
+    ENDIF
+  ENDFUNCTION at_string_slice
+!
+!-------------------------------------------------------------------------------
+!> @brief Utility function takes a string and converts all upper case letters to
+!> lower case letters.
+!> @param word input is a string, output has all lower case letters
+!>
+    PURE FUNCTION toLower_string(word) RESULT(upper_str)
+      CLASS(StringType),INTENT(IN) :: word
+      INTEGER(SIK) :: i
+      !
+      TYPE(StringType) :: upper_str
+      upper_str = word
+      DO i=1,LEN(word)
+        IF('A' <= word%s(i:i) .AND. word%s(i:i) <= 'Z') &
+          upper_str%s(i:i)=ACHAR(IACHAR(word%s(i:i))+32)
+      ENDDO
+    ENDFUNCTION toLower_string
+!
+!-------------------------------------------------------------------------------
+!> @brief Utility function takes a string and converts all lower case letters to
+!> upper case letters.
+!> @param word input is a string, output has all upper case letters
+!>
+    PURE FUNCTION toupper_string(word) RESULT(lower_str)
+      CLASS(StringType),INTENT(IN) :: word
+      INTEGER(SIK) :: i
+      !
+      TYPE(StringType) :: lower_str
+      lower_str = word
+      DO i=1,LEN(word)
+        IF('a' <= word%s(i:i) .AND. word%s(i:i) <= 'z') &
+          lower_str%s(i:i)=ACHAR(IACHAR(word%s(i:i))-32)
+      ENDDO
+    ENDFUNCTION toupper_string
+!
+!-------------------------------------------------------------------------------
 !> @brief Returns the length of the string including all whitespace.
-!> @param thisStr the string object
+!> @param this the string object
 !> @returns n the length of the string
 !>
-!> The intent is that this behaves exactly the same way as the intrinsic
-!> function @c LEN does for character variables.
-!>
-    ELEMENTAL FUNCTION LEN_StringType(thisStr) RESULT(n)
-      CLASS(StringType),INTENT(IN) :: thisStr
+!> Should behave exactly as it would if called directly on a character array
+    ELEMENTAL FUNCTION LEN_StringType(this) RESULT(n)
+      CLASS(StringType),INTENT(IN) :: this
       INTEGER(SIK) :: n
-      n=thisStr%n
+      n=LEN(this%s)
```

Comment:
why get rid of %n?

Reply:
replaced its usage with `LEN()`. I think it was only around because the characters were stored as `CHARACTER(LEN=1), ALLOCATABLE ::: s(:)` which isn't always consistent with the string that `StringType` replaced. So once I changed the internal representation of `s`, there was no need to have an `n`, since you could just `LEN(StringType_instance)`

Reply:
same with `ntrim` just `LEN_TRIM`